File: composeApp/src/androidMain/kotlin/dev/tcode/thinmpk/player/MusicPlayerImpl.kt
package dev.tcode.thinmpk.player

import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.os.IBinder
import dev.tcode.thinmpk.config.RepeatState
import dev.tcode.thinmpk.model.SongModel

class MusicPlayerImpl(private val context: Context) : MusicPlayer {
    private var musicService: MusicService? = null
    private lateinit var connection: ServiceConnection
    private var isServiceBinding = false
    private var bound = false
    private val listeners: MutableList<MusicPlayerListener> = mutableListOf()
    private val serviceListener = object : MusicPlayerListener {
        override fun onChange() {
            listeners.forEach { it.onChange() }
        }

        override fun onError() {
            listeners.forEach { it.onError() }
        }
    }

    override fun start(songs: List<SongModel>, index: Int) {
        if (isServiceBinding) return

        if (!MusicService.isServiceRunning) {
            context.startForegroundService(Intent(context, MusicService::class.java))
            bindService { musicService?.start(songs, index) }
            return
        }

        if (!bound) {
            bindService { musicService?.start(songs, index) }
            return
        }

        musicService?.start(songs, index)
    }

    override fun play() {
        musicService?.play()
    }

    override fun pause() {
        musicService?.pause()
    }

    override fun prev() {
        musicService?.prev()
    }

    override fun next() {
        musicService?.next()
    }

    override fun seekTo(ms: Long) {
        musicService?.seekTo(ms)
    }

    override fun isPlaying(): Boolean {
        return musicService?.isPlaying() == true
    }

    override fun getCurrentSong(): SongModel? {
        return musicService?.getCurrentSong()
    }

    override fun getCurrentPosition(): Long {
        return musicService?.getCurrentPosition() ?: 0
    }

    override fun getRepeat(): RepeatState {
        return musicService?.getRepeat() ?: RepeatState.OFF
    }

    override fun changeRepeat() {
        musicService?.changeRepeat()
    }

    override fun getShuffle(): Boolean {
        return musicService?.getShuffle() ?: false
    }

    override fun changeShuffle() {
        musicService?.changeShuffle()
    }

    override fun addEventListener(listener: MusicPlayerListener) {
        listeners.add(listener)
    }

    override fun removeEventListener(listener: MusicPlayerListener) {
        listeners.remove(listener)
    }

    private fun bindService(callback: () -> Unit = {}) {
        if (isServiceBinding || bound) return

        isServiceBinding = true
        connection = createConnection(callback)
        context.bindService(
            Intent(context, MusicService::class.java),
            connection,
            Context.BIND_AUTO_CREATE
        )
    }

    private fun createConnection(callback: () -> Unit = {}): ServiceConnection {
        return object : ServiceConnection {
            override fun onServiceConnected(name: ComponentName, service: IBinder) {
                val binder = service as MusicService.MusicBinder
                musicService = binder.getService()
                musicService!!.addEventListener(serviceListener)
                callback()
                isServiceBinding = false
                bound = true
            }

            override fun onServiceDisconnected(name: ComponentName) {
                musicService = null
                bound = false
            }
        }
    }
}

File: iosApp/iosApp/Player/MusicPlayerImpl.swift
import MediaPlayer
import ComposeApp

class MusicPlayerImpl: ComposeApp.MusicPlayer {
    private let PREV_SECOND: Double = 3
    private let player: MPMusicPlayerController
    private let playerConfig: PlayerConfig
    private var listeners: [MusicPlayerListener] = []
    private var playingList: [SongModel] = []
    private var currentRepeat: RepeatState = .off
    private var currentShuffle: Bool = false

    init() {
        playerConfig = PlayerConfig()
        player = MPMusicPlayerController.applicationMusicPlayer
        player.repeatMode = playerConfig.getRepeat()
        player.shuffleMode = playerConfig.getShuffle()
        currentRepeat = Self.toRepeatState(mode: player.repeatMode)
        currentShuffle = player.shuffleMode == .songs
        addObserver()
        player.beginGeneratingPlaybackNotifications()
    }

    func start(songs: [SongModel], index: Int32) {
        if player.playbackState == .playing {
            player.stop()
        }

        playingList = songs

        let query = MPMediaQuery.songs()
        query.addFilterPredicate(
            MPMediaPropertyPredicate(value: false, forProperty: MPMediaItemPropertyIsCloudItem)
        )

        guard let allItems = query.items else { return }

        let songIds = songs.map { $0.id }
        let mediaItems = songIds.compactMap { songId in
            allItems.first { String($0.persistentID) == songId }
        }

        guard !mediaItems.isEmpty else { return }

        let collection = MPMediaItemCollection(items: mediaItems)
        let descriptor = MPMusicPlayerMediaItemQueueDescriptor(itemCollection: collection)

        if Int(index) < mediaItems.count {
            descriptor.startItem = mediaItems[Int(index)]
        }

        player.setQueue(with: descriptor)
        play()
    }

    func play() {
        player.play()
    }

    func pause() {
        player.pause()
    }

    func prev() {
        if player.currentPlaybackTime <= PREV_SECOND {
            player.skipToPreviousItem()
        } else {
            player.skipToBeginning()
        }
    }

    func next() {
        player.skipToNextItem()
    }

    func seekTo(ms: Int64) {
        player.currentPlaybackTime = Double(ms) / 1000.0
    }

    func isPlaying() -> Bool {
        return player.playbackState == .playing
    }

    func getCurrentSong() -> SongModel? {
        guard let nowPlaying = player.nowPlayingItem else { return nil }
        let persistentId = String(nowPlaying.persistentID)

        return playingList.first { $0.id == persistentId }
    }

    func getCurrentPosition() -> Int64 {
        return Int64(player.currentPlaybackTime * 1000)
    }

    func getRepeat() -> RepeatState {
        return currentRepeat
    }

    func changeRepeat() {
        player.repeatMode = player.repeatMode == .none ? .all
            : player.repeatMode == .all ? .one
            : .none
        currentRepeat = Self.toRepeatState(mode: player.repeatMode)
        playerConfig.setRepeat(value: player.repeatMode)
        onChange()
    }

    func getShuffle() -> Bool {
        return currentShuffle
    }

    func changeShuffle() {
        player.shuffleMode = player.shuffleMode == .off ? .songs : .off
        currentShuffle = player.shuffleMode == .songs
        playerConfig.setShuffle(value: player.shuffleMode)
        onChange()
    }

    func addEventListener(listener: MusicPlayerListener) {
        listeners.append(listener)
    }

    func removeEventListener(listener: MusicPlayerListener) {
        listeners.removeAll { $0 === listener }
    }

    private func onChange() {
        listeners.forEach { $0.onChange() }
    }

    private func addObserver() {
        NotificationCenter.default.addObserver(
            forName: .MPMusicPlayerControllerNowPlayingItemDidChange,
            object: player,
            queue: .main
        ) { [weak self] _ in
            self?.onChange()
        }

        NotificationCenter.default.addObserver(
            forName: .MPMusicPlayerControllerPlaybackStateDidChange,
            object: player,
            queue: .main
        ) { [weak self] _ in
            self?.onChange()
        }
    }

    private static func toRepeatState(mode: MPMusicRepeatMode) -> RepeatState {
        switch mode {
        case .one: return .one
        case .all: return .all
        default: return .off
        }
    }

    deinit {
        NotificationCenter.default.removeObserver(self)
        player.endGeneratingPlaybackNotifications()
    }
}
